hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapProxySupport.java

```diff
@@ -16,54 +16,268 @@
+                    //FIXME: implement finalizePutAll(map);
+                    if (!resultFuture.isDone()) {
+                        resultFuture.complete(null);
+                    }
+                }
+            };
+            for (Map.Entry<Address, List<Integer>> entry : memberPartitionsMap.entrySet()) {
+                invokePutAllOperation(entry.getKey(), entry.getValue(), entriesPerPartition).whenCompleteAsync(callback);
+            }
+            // if executing in sync mode, block for the responses
+            if (future == null) {
+                resultFuture.get();
+            }
+        } catch (Throwable e) {
+            throw rethrow(e);
+        }
+    }
+
+    //NB: this method is generally copied from MapProxySupport#invokePutAllOperation
+    private InternalCompletableFuture<Void> invokePutAllOperation(
+            Address address,
+            List<Integer> memberPartitions,
+            MapEntries[] entriesPerPartition
+    ) {
+        int size = memberPartitions.size();
+        int[] partitions = new int[size];
+        int index = 0;
+        for (Integer partitionId : memberPartitions) {
+            if (entriesPerPartition[partitionId] != null) {
+                partitions[index++] = partitionId;
+            }
+        }
+        if (index == 0) {
+            return newCompletedFuture(null);
+        }
+        // trim partition array to real size
+        if (index < size) {
+            partitions = Arrays.copyOf(partitions, index);
+            size = index;
+        }
+
+        index = 0;
+        MapEntries[] entries = new MapEntries[size];
+        long totalSize = 0;
+        for (int partitionId : partitions) {
+            int batchSize = entriesPerPartition[partitionId].size();
+            //FIXME: implement batching
+            //assert (putAllBatchSize == 0 || batchSize <= putAllBatchSize);
+            entries[index++] = entriesPerPartition[partitionId];
+            totalSize += batchSize;
+            entriesPerPartition[partitionId] = null;
+        }
+        if (totalSize == 0) {
+            return newCompletedFuture(null);
+        }
+
+        OperationFactory factory = new MultiMapPutAllOperationFactory(name, partitions, entries);
+        long startTimeNanos = System.nanoTime();
+        CompletableFuture<Map<Integer, Object>> future =
+                operationService.invokeOnPartitionsAsync(MultiMapService.SERVICE_NAME, factory,
+                        singletonMap(address, asIntegerList(partitions)));
+        InternalCompletableFuture<Void> resultFuture = new InternalCompletableFuture<>();
+        long finalTotalSize = totalSize;
+        future.whenCompleteAsync((response, t) -> {
+            //TODO: implement putAllVisitSerializedKeys
+            //putAllVisitSerializedKeys(entries);
+
+            if (t == null) {
+                //TODO: verify stats
+                getService().getLocalMultiMapStatsImpl(name).incrementPutLatencyNanos(finalTotalSize,
```

Comment:
consider adding test for stats of `putAllAsync` in `MultiMapContainerStatisticsTest`